File: packages/i18nify-go/modules/locale/resolve_locale.go
package locale

import (
	"os"
	"strings"
)

// ResolveLocaleOptions configures the locale resolution priority chain.
type ResolveLocaleOptions struct {
	// Locale is an explicit locale preference — the "options" layer.
	// Corresponds to the options layer in the client-side priority chain.
	Locale string

	// FallbackLocale is returned when no higher-priority signal resolves.
	// Defaults to "en" when empty.
	FallbackLocale string
}

// ResolveLocale resolves a BCP 47 locale string by walking a priority chain.
//
// The server-side chain maps the four client-side layers as follows:
//
//	Client-side    → Go equivalent
//	───────────────────────────────────────────────────────────────
//	storage        → (not applicable — no localStorage in Go)
//	options        → ResolveLocaleOptions.Locale
//	browser        → LC_ALL / LC_MESSAGES / LANG environment variables
//	fallback       → ResolveLocaleOptions.FallbackLocale (default: "en")
//
// POSIX charset suffixes are stripped: "en_US.UTF-8" → "en_US".
// "C" and "POSIX" locale values are treated as absent.
// The function always returns a non-empty string — it never errors.
//
// Example:
//
//	ResolveLocale(ResolveLocaleOptions{Locale: "fr-FR"})
//	// → "fr-FR"
//
//	ResolveLocale(ResolveLocaleOptions{FallbackLocale: "ja"})
//	// → value of LANG env, or "ja" if LANG is unset
//
//	ResolveLocale(ResolveLocaleOptions{})
//	// → value of LANG env, or "en" if LANG is unset
func ResolveLocale(opts ResolveLocaleOptions) string {
	// Priority 1: explicit locale option (corresponds to "options" layer)
	if trimmed := strings.TrimSpace(opts.Locale); trimmed != "" {
		return trimmed
	}

	// Priority 2: OS locale environment variables (corresponds to "browser" layer)
	for _, envVar := range []string{"LC_ALL", "LC_MESSAGES", "LANG"} {
		val := strings.TrimSpace(os.Getenv(envVar))
		if val == "" || val == "C" || val == "POSIX" {
			continue
		}
		// Strip charset suffix: "en_US.UTF-8" → "en_US"
		if dot := strings.IndexByte(val, '.'); dot != -1 {
			val = val[:dot]
		}
		return val
	}

	// Priority 3: fallback
	if fallback := strings.TrimSpace(opts.FallbackLocale); fallback != "" {
		return fallback
	}
	return "en"
}

File: packages/i18nify-js/src/modules/geo/resolveLocale.ts
import { withErrorBoundary } from '../../common/errorBoundary';

type ResolveLocaleOptions = {
  /** localStorage key to check. Default: 'i18nify-locale' */
  storageKey?: string;
  /** Explicit locale preference — the "options" layer. */
  locale?: string;
  /** Returned when all higher-priority signals are absent. Default: 'en' */
  fallbackLocale?: string;
};

/**
 * Resolves a BCP 47 locale string by walking a four-level priority chain:
 *
 *   1. **storage** — `localStorage.getItem(storageKey)` (client only)
 *   2. **options** — the explicit `locale` field
 *   3. **browser** — `navigator.languages[0]` / `navigator.language` (client only)
 *   4. **fallback** — `fallbackLocale` (default `'en'`)
 *
 * Each level is skipped if its value is empty/whitespace or unavailable.
 * The function is synchronous and never throws — it always returns a string.
 *
 * @example
 * // localStorage has 'i18nify-locale' = 'fr-FR'
 * resolveLocale()                                  // → 'fr-FR'
 *
 * // no storage, explicit locale passed
 * resolveLocale({ locale: 'de-DE' })               // → 'de-DE'
 *
 * // no storage, no explicit locale, browser says 'pt-BR'
 * resolveLocale()                                  // → 'pt-BR'
 *
 * // SSR / no browser APIs, custom fallback
 * resolveLocale({ fallbackLocale: 'en-US' })       // → 'en-US'
 */
const resolveLocale = (options: ResolveLocaleOptions = {}): string => {
  const {
    storageKey = 'i18nify-locale',
    locale,
    fallbackLocale = 'en',
  } = options;

  // Priority 1: storage (localStorage, client environments only)
  if (typeof window !== 'undefined' && window.localStorage) {
    const stored = window.localStorage.getItem(storageKey);
    if (stored?.trim()) return stored.trim();
  }

  // Priority 2: options (explicitly passed locale)
  if (locale?.trim()) return locale.trim();

  // Priority 3: browser (navigator.languages / navigator.language)
  if (typeof window !== 'undefined' && window.navigator) {
    const browserLocale =
      window.navigator.languages?.[0] || window.navigator.language;
    if (browserLocale?.trim()) return browserLocale.trim();
  }

  // Priority 4: fallback
  return fallbackLocale?.trim() || 'en';
};

export default withErrorBoundary<typeof resolveLocale>(resolveLocale);
